### PR TITLE
is0502: add test for staging of transport_file

### PR DESCRIPTION
**Pull request by andrewbonney**
A partial resolution to #341

### CODE REVIEW

**Comment by andrewbonney** [comment]
This could be extended later to test 2022-7 and changing transport_params in the same request, but hopefully this will catch the majority of issues in this area.

File: nmostesting/suites/IS0502Test.py [hunk 1]
```diff
@@ -1052,6 +1055,99 @@ def test_17(self, test):
...
+        """Receiver correctly translates SDP file attributes into transport_params"""
+
+        self.do_test_node_api_v1_2(test)
+
+        valid, result = self.get_is04_resources("receivers")
+        if not valid:
+            return test.FAIL(result)
+
+        if len(self.is04_resources["receivers"]) == 0:
+            return test.UNCLEAR("Could not find any IS-04 Senders to test")
```
Comment (garethsb-sony) [inline nmostesting/suites/IS0502Test.py]: Receivers :-)

Comment (andrewbonney) [reply nmostesting/suites/IS0502Test.py]: Fail

**Comment by garethsb-sony** [comment]
At some point, could also try to extend test to support unicast destination address (setting `interface_ip` rather than `multicast_ip`).

**Comment by garethsb-sony** [comment]
In the multicast case being tested, where `"multicast_ip"` is being set from the destination address, I wonder if `"interface_ip"` ought to be required to be reset to `"auto"`?

In general if you're just tweaking `"transport_params"` you wouldn't want any you hadn't asked for to change, but for an SDP file the `"transport_params"` are somewhat out of your control. There are cases with things like `"rtp_enabled`" for example where they change but don't relate directly to a param in the SDP.

I had a concern that having a random left-over `"interface_ip"` value could potentially cause trouble with receiving a RTP multicast stream specified by SDP?

**Comment by simonlo-sony** [comment]
For IS0502Test 
1. test_14 - rtpmap test condition for metadata and mux are incorrect
2. test_15 - according from the scheme grain_rate is allowed not be presented in the source

```
diff --git a/nmostesting/suites/IS0502Test.py b/nmostesting/suites/IS0502Test.py
index 593df4e..149972b 100644
--- a/nmostesting/suites/IS0502Test.py
+++ b/nmostesting/suites/IS0502Test.py
@@ -684,7 +684,7 @@ class IS0502Test(GenericTest):
                             if flow["media_type"] == "video/smpte291" and not rtpmap:
                                 return test.FAIL("a=rtpmap clock rate does not match expected rate for Flow media "
                                                  "type {} and Sender {}".format(flow["media_type"], resource["id"]))
-                            elif rtpmap:
+                            elif not rtpmap:
                                 return test.FAIL("a=rtpmap specifies a different media_type to the Flow for Sender {}"
                                                  .format(resource["id"]))
                         elif source["format"] == "urn:x-nmos:format:mux":
@@ -692,7 +692,7 @@ class IS0502Test(GenericTest):
                             if flow["media_type"] == "video/SMPTE2022-6" and not rtpmap:
                                 return test.FAIL("a=rtpmap clock rate does not match expected rate for Flow media "
                                                  "type {} and Sender {}".format(flow["media_type"], resource["id"]))
-                            elif rtpmap:
+                            elif not rtpmap:
                                 return test.FAIL("a=rtpmap specifies a different media_type to the Flow for Sender {}"
                                                  .format(resource["id"]))
         except KeyError as ex:
@@ -821,7 +821,7 @@ class IS0502Test(GenericTest):
                                     return test.FAIL("Segmented parameter for Sender {} incorrectly includes an '='"
                                                      .format(resource["id"]))
                             elif param_components[0] == "exactframerate":  # ref: ST.2110-20
-                                if "grain_rate" not in source:
+                                if "grain_rate" in source:
                                     return test.FAIL("No grain_rate found for Source {} associated with Sender {}"
                                                      .format(source["id"], resource["id"]))
                                 if "grain_rate" in flow:
```

**Comment by andrewbonney** [comment]
Hi Simon. I think these issues relate to different tests which were merged in different pull requests. The test_14 issue may need reporting as a bug, but I'll investigate here and may be able to do a quick fix. For test_15 the current behaviour is intentional as it would be unusual for most Sources not to be periodic. Whilst the schemas don't require the 'grain_rate', they do state that it is mandatory for periodic Sources.

**Comment by garethsb-sony** [comment]
@andrewbonney, is the concern expressed above - https://github.com/AMWA-TV/nmos-testing/pull/442#issuecomment-567886138 - valid?

**Comment by andrewbonney** [comment]
Possibly. I'm wondering if any required changes to this parameter ought to be handled automatically though given an SDP file is being passed in, which has an impact on all other `transport_params`. Admittedly I don't think this is explicitly called out anywhere.

I don't think it would be problematic to include the additional 'auto' in the testing tool request, but whether controllers should do this as a matter of course is probably another matter. It might be considered a breaking behavioural change to require that at this stage without a new version.

**Comment by garethsb-sony** [comment]
Ah, yes, OK, to be clear:

If the Receiver PATCH includes an SDP `"transport_file"`, then it's the Receiver's job to set both its `"interface_ip"` and `"multicast_ip"` based on that file (one of them to the address and the other to `"auto"` or `null` respectively).

On the other hand, when a controller is setting `"transport_params"` directly, without an SDP file, it is recommended that the controller sets those parameters together, because the Receiver will not automatically update the other if only one of them is patched.

**Pull request merged**